Split posts and comments into separate Elasticsearch indices

mbertrand opened the pull request:
#### Pre-Flight checklist
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #1337 

#### What's this PR do?
- Creates (and recreates) separate indices for posts and comments
- Creates an alias to allow searching both indices simultaneously
- Defines separate mappings for posts and comments
- Adds `channel_name` to both post and comment mappings

#### How should this be manually tested?
- Run `python manage.py recreate_index`
- After it's completed, posts should be found in a `discussions_local_post_default` index, comments in a `discussions_local_comment_default` index, and there should be an alias `discussions_local_all_default`.  The original `discussions_local_default` index from master branch will still exist unless you manually delete it or never created it in the first place.
- If you post the following to `http://localhost:9101/discussions_local_all_default/_search/`, you should get a list of posts back:
  ```javascript
  {
    "query": {
      "term": {
        "object_type": "post"
      }
    }
  }
  ```
- If you post the above but replace `post` with `comment`, you should get a list of comments back.

- Create a new post, then add some comments.  The new post should be added to the post index only, the comments should be added to the comment index only.  
- 'Remove' a post that has comments, then search for the records in Elasticsearch.  The comments should have `parent_post_removed: true`

noisecapella commented on search/indexing_api.py:
```diff
@@ -44,25 +43,35 @@
+        object_type(str): The type of document (post, comment)
     """
+    if not object_type or object_type not in VALID_OBJECT_TYPES:
+        raise TypeError('A valid object type must be specified when clearing and creating an index')
```
This is pretty nitpicky but we should use `ValueError` here. According to [the docs](https://docs.python.org/3/library/exceptions.html#TypeError) `TypeError` is only if the Python types don't match with each other

mbertrand replied:
✅

noisecapella commented on search/indexing_api.py:
```diff
@@ -263,8 +292,22 @@ def index_post_with_comments(post_id):
+            chunk_size=settings.ELASTICSEARCH_INDEXING_CHUNK_SIZE,
+        )
+        if len(errors) > 0:
+            raise ReindexException("Error during bulk insert: {errors}".format(
```
Can you change the message here and above to say whether it was comment or post we're indexing?

mbertrand replied:
✅

noisecapella commented on search/indexing_api.py:
```diff
@@ -263,8 +292,22 @@ def index_post_with_comments(post_id):
+
+def create_backing_index(object_type):
     """
     Start the reindexing process by creating a new backing index and pointing the reindex alias toward it
```
Can you add `object_type` to the docstring?

mbertrand replied:
✅

noisecapella commented on search/indexing_api.py:
```diff
@@ -291,7 +334,7 @@ def create_backing_index():
-def switch_indices(backing_index):
+def switch_indices(backing_index, object_type):
     """
     Switch the default index to point to the backing index, and delete the reindex alias
```
Can you add `object_type` to the docstring?

mbertrand replied:
✅

noisecapella commented on search/serializers_test.py:
```diff
@@ -133,8 +132,7 @@ def test_serialize_bulk_post_and_comments(mocker, patched_base_post_serializer,
     post_mock = mocker.MagicMock(comments=[outer_comment_mock])
-    assert list(serialize_bulk_post_and_comments(post_mock)) == [
-        serialize_post_for_bulk(post_mock),
+    assert list(serialize_bulk_comments(post_mock)) == [
```
Can you change the name of this test? Do we have a test for serializing posts?

mbertrand replied:
✅ , `test_serialize_post_for_bulk`

noisecapella commented on search/tasks.py:
```diff
@@ -159,13 +158,13 @@ def start_recreate_index(self):
 
 @app.task
-def finish_recreate_index(results, backing_index):
+def finish_recreate_index(results, backing_indices):
```
Can you update the docstring?

mbertrand replied:
✅

mbertrand commented on search/connection.py:
```diff
@@ -52,56 +53,69 @@ def get_conn(*, verify=True):
     conn = get_conn(verify=False)
     return [
-        alias for alias in [get_default_alias_name(), get_reindexing_alias_name()]
+        alias for x in [
```
✅

mbertrand commented on search/indexing_api.py:
```diff
@@ -44,25 +43,35 @@
+}
+
+MAPPING = {
+    POST_TYPE: dict(BASE_CONTENT_TYPE, **{
```
✅

mbertrand commented on search/indexing_api.py:
```diff
@@ -44,25 +43,35 @@
         skip_mapping (bool): If true, don't set any mapping
+        object_type(str): The type of document (post, comment)
     """
+    if not object_type or object_type not in VALID_OBJECT_TYPES:
```
✅

mbertrand commented:
@noisecapella I added `channel_name` to the mappings, and fixed a bug I noticed where `channel_title` was getting mixed up with `channel_name` in the comments index.

noisecapella commented:
Thanks, functionality is looking great. Looking over code now
